experiments/yarn/src/main/scala/org/apache/gearpump/experiments/yarn/master/YarnApplicationMaster.scala

```diff
@@ -0,0 +1,315 @@
+    containers.foreach(container => {
+      val port = appConfig.getEnv(GEARPUMPMASTER_PORT).toInt
+      val masterCommand = MasterContainerCommand(appConfig, masterAddr.get)
+      launchCommand(container, masterCommand.getCommand)
+      masterContainers += container.getId ->(container.getNodeId.getHost, port)
+    })
+  }
+
+  private def launchWorkerContainers(containers: List[Container]) {
+    containers.foreach(container => {
+      val workerHost = container.getNodeId.getHost
+      val workerCommand = WorkerContainerCommand(appConfig, masterAddr.get, workerHost).getCommand
+      launchCommand(container, workerCommand)
+    })
+  }
+
+  private def launchCommand(container: Container, command: String) {
+    LOG.info(s"Launching containter: containerId :  ${container.getId}, host ip : ${container.getNodeId.getHost}")
+    LOG.info("Launching command : " + command)
+    val containerContext = containerLaunchContextFactory(command)
+    LOG.info(s"NodeManagerClient : $nodeManagerClient, container : $container, containerContext $containerContext")
+    nodeManagerClient.startContainerAsync(container, containerContext)
+  }
+
+  private def requestWorkerContainers(): Unit = {
+    (1 to appConfig.getEnv(WORKER_CONTAINERS).toInt).foreach(requestId => {
+      resourceManagerClient ! ContainersRequest(appConfig.getEnv(WORKER_MEMORY).toInt, appConfig.getEnv(WORKER_VCORES).toInt)
+    })
+  }
+
+  private def requestMoreContainers(count: Int): Unit = {
+    (1 to count).foreach((i) => resourceManagerClient ! ContainersRequest(appConfig.getEnv(WORKER_MEMORY).toInt, appConfig.getEnv(WORKER_VCORES).toInt))
+  }
+
+  private def requestMasterContainers(registrationResponse: RegisterApplicationMasterResponse) = {
+    //TODO remove that or do something more with that
+    val previousContainersCount = registrationResponse.getContainersFromPreviousAttempts.size
+
+    LOG.info(s"Previous container count : $previousContainersCount")
+    if (previousContainersCount > 0) {
+      LOG.warn("Previous container count > 0, can't do anything with it")
+    }
+
+    LOG.info(s"GEARPUMPMASTER_CONTAINERS: ${appConfig.getEnv(GEARPUMPMASTER_CONTAINERS).toInt}")
+    (1 to appConfig.getEnv(GEARPUMPMASTER_CONTAINERS).toInt).foreach(requestId => {
+      resourceManagerClient ! ContainersRequest(appConfig.getEnv(GEARPUMPMASTER_MEMORY).toInt, appConfig.getEnv(GEARPUMPMASTER_VCORES).toInt)
+    })
+
+  }
+
+  private def cleanUp(state: RMTerminalState): Unit = {
+    LOG.info("Application completed. Stopping running containers")
+    nodeManagerClient.stop()
+
+    state match {
+      case RMError(throwable, stats) =>
+        LOG.info("Failed", throwable.getMessage)
+        val message = s"Failed. total=${appConfig.getEnv(WORKER_CONTAINERS).toInt}, completed=${stats.completed}, allocated=${stats.allocated}, failed=${stats.failed}"
+        resourceManagerClient ! AMStatusMessage(FinalApplicationStatus.FAILED, message, null)
```

Comment:
Removed AMStatusMessage. Logging of terminal state and unregistering application master is now done in ResourceManagerClient.